lib/resumption/handover.go

```diff
@@ -0,0 +1,112 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/netip"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/multiplexer"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func (r *SSHServerWrapper) attemptHandover(conn *multiplexer.Conn, token resumptionToken) {
+	handoverConn, err := r.dialHandover(token)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			r.log.Debug("Resumable connection not found or already deleted.")
+			_, _ = conn.Write([]byte{notFoundServerExchangeTag})
+			return
+		}
+		r.log.WithError(err).Error("Error while connecting to handover socket.")
+		return
+	}
+	defer handoverConn.Close()
+
+	var remoteIP netip.Addr
+	if t, _ := conn.RemoteAddr().(*net.TCPAddr); t != nil {
+		remoteIP, _ = netip.AddrFromSlice(t.IP)
+	}
+	remoteIP16 := remoteIP.As16()
+
+	if _, err := handoverConn.Write(remoteIP16[:]); err != nil {
+		if !utils.IsOKNetworkError(err) {
+			r.log.WithError(err).Error("Error while forwarding remote address to handover socket.")
+		}
+		return
+	}
+
+	r.log.Debug("Forwarding resuming connection to handover socket.")
+	_ = utils.ProxyConn(context.Background(), conn, handoverConn)
+}
+
+func (r *SSHServerWrapper) setupHandoverListener(ctx context.Context, token resumptionToken, entry *connEntry) error {
+	l, err := r.listenHandover(token)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go r.runHandoverListener(l, entry)
+	context.AfterFunc(ctx, func() { _ = l.Close() })
+
+	return nil
+}
+
+func (r *SSHServerWrapper) runHandoverListener(l net.Listener, entry *connEntry) {
+	defer l.Close()
+
+	var tempDelay time.Duration
+	for {
+		c, err := l.Accept()
+		if err == nil {
+			tempDelay = 0
+			go r.handleHandoverConnection(c, entry)
+			continue
+		}
+
+		if tempErr, ok := err.(interface{ Temporary() bool }); !ok || !tempErr.Temporary() {
+			if !utils.IsOKNetworkError(err) {
+				r.log.WithError(err).Warn("Accept error in handover listener.")
+			}
+			return
+		}
+
+		tempDelay = max(5*time.Millisecond, min(2*tempDelay, time.Second))
+		r.log.WithError(err).WithField("delay", tempDelay).Warn("Temporary accept error in handover listener, continuing after delay.")
+		time.Sleep(tempDelay)
```

Comment:
Is this a common occurrence? Worth a comment explaining what's going on here.

Reply:
Would you mind linking to that or something? I'm not seeing many other instances of the `Temporary()` bit.

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/net/http/server.go;l=3254-3286 in go 1.22; the `Temporary()` method on `net.Error` (which is what `(*http.Server).Serve` type asserts on) was deprecated without any sensible replacement for it, but it's still probably better than only checking for `ErrClosed` and blindly retrying errors that will never fix themselves (like if the network card disappears, or something).